pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,633 @@
+	// mu controls access to the electionCandidateTracker map
+	mu sync.RWMutex
+}
+
+type election struct {
+	leaderLeaseID leaderLeaseID
+	electionStart time.Time
+}
+
+type candidateStatus struct {
+	lock       sync.Mutex
+	candidates map[string]bool
+	done       chan bool
+	total      int
+}
+
+func (c *candidateStatus) Ack(name string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.candidates[name] = true
+	if len(c.candidates) == c.total {
+		close(c.done)
+	}
+}
+
+func (c *Controller) initializeCT(leaderLeaseID leaderLeaseID, total int) *candidateStatus {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.electionCandidateTracker[leaderLeaseID] = &candidateStatus{
+		candidates: make(map[string]bool),
+		done:       make(chan bool),
+		total:      total,
+	}
+	return c.electionCandidateTracker[leaderLeaseID]
+}
+
+func (c *Controller) getCT(leaderLeaseID leaderLeaseID) *candidateStatus {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	ct, ok := c.electionCandidateTracker[leaderLeaseID]
+	if ok {
+		return ct
+	}
+	return nil
+
+}
+
+type leaderLeaseID struct {
+	namespace, name string
+}
+
+func parseLeaderLeaseID(namespace, id string) (leaderLeaseID, error) {
+	return leaderLeaseID{namespace, id}, nil
+}
+
+func (l leaderLeaseID) String() string {
+	return l.namespace + "/" + l.name
+}
+
+func (c *Controller) Sync(ctx context.Context) {
+	if !cache.WaitForNamedCacheSync(controllerName, ctx.Done(), c.leaseSynced) {
+		return
+	}
+	if !cache.WaitForNamedCacheSync(controllerName, ctx.Done(), c.leaseCandidateSynced) {
+		return
+	}
+
+}
+
+func (c *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+
+	defer c.leaseQueue.ShutDown()
+	klog.Infof("Workers: %d", workers)
+	for i := 0; i < workers; i++ {
+		klog.Infof("Starting worker")
+		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func NewGenerator(ctx context.Context, config *rest.Config, leaseInformer coordinationv1.LeaseInformer, leaseCandidateInformer coordinationv1alpha1.LeaseCandidateInformer, leaseClient coordinationv1client.CoordinationV1Interface, leaseCandidateClient coordinationv1alpha1client.CoordinationV1alpha1Interface) (*Controller, error) {
+	var cancel context.CancelFunc
+
+	callbacks := leaderelection.LeaderCallbacks{
+		OnStartedLeading: func(ctx context.Context) {
+			var genCtx context.Context
+			controller, err := NewController(leaseInformer, leaseCandidateInformer, leaseClient, leaseCandidateClient)
+			if err != nil {
+				panic(err)
+			}
+			genCtx, cancel = context.WithCancel(ctx)
+			controller.Sync(genCtx)
+			controller.Run(genCtx, 2)
+		},
+		OnStoppedLeading: func() {
+			cancel()
+		},
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	rl, err := resourcelock.NewFromKubeconfig(
+		"leases",
+		"kube-system",
+		"leader-election-controller",
+		resourcelock.ResourceLockConfig{
+			Identity: hostname + "_" + string(uuid.NewUUID()),
```

Comment:
we already have an apiserver identity, wonder if we can reuse it here so we have consistency instead of creating new identities for different features

Reply:
apiserver identity uses the convention of `apiserverID + "_" + uuid`. Leases use the convention of `hostname + "_" + uuid`. Wouldn't it be better to adhere to how other leases are named?